File: ViewModels/BudgetViewModel.swift
import Foundation
import SwiftUI

class BudgetViewModel: ObservableObject {

    @Published var budgets: [Budget] = DefaultCategories.all.map {
        Budget(
            category: $0,
            limit: 0,
            spent: 0
        )
    }

    // MARK: - Add Budget
    func addBudget(_ budget: Budget) {
        budgets.append(budget)
    }

    // MARK: - Set or Update Budget
    func setBudget(for category: Category, limit: Double) {
        if let index = budgets.firstIndex(where: { $0.category.id == category.id }) {
            budgets[index].limit = limit
        } else {
            let newBudget = Budget(
                category: category,
                limit: limit,
                spent: 0
            )
            budgets.append(newBudget)
        }
    }

    // MARK: - Budget Lookup
    func budget(for category: Category) -> Budget? {
        budgets.first { $0.category.id == category.id }
    }

    // MARK: - Update Budgets From Transactions
    func updateBudgets(
        transactions: [Transaction],
        startDate: Date,
        endDate: Date
    ) {
        for index in budgets.indices {
            let category = budgets[index].category

            let spentAmount = transactions
                .filter {
                    !$0.isIncome &&
                    $0.category.id == category.id &&
                    $0.date >= startDate &&
                    $0.date <= endDate
                }
                .reduce(0) { $0 + $1.amount }

            budgets[index].spent = spentAmount
        }
    }

    // MARK: - Reset Budgets
    func resetBudgets(carryOverEnabled: Bool) {
        for index in budgets.indices {
            if carryOverEnabled {
                let remaining = budgets[index].remaining
                budgets[index].spent = remaining > 0 ? -remaining : 0
            } else {
                budgets[index].spent = 0
            }
        }
    }

    // MARK: - Totals
    var totalLimit: Double {
        budgets.reduce(0) { $0 + $1.limit }
    }

    var totalSpent: Double {
        budgets.reduce(0) { $0 + $1.spent }
    }

    var totalRemaining: Double {
        budgets.reduce(0) { $0 + $1.remaining }
    }
}

File: ledgerly-android/app/src/main/java/com/ledgerly/viewmodel/BudgetViewModel.kt
package com.ledgerly.viewmodel

import android.app.Application
import androidx.compose.runtime.mutableStateListOf
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.ledgerly.data.database.DatabaseProvider
import com.ledgerly.data.database.toEntity
import com.ledgerly.data.database.toModel
import com.ledgerly.data.models.Budget
import com.ledgerly.data.models.Category
import com.ledgerly.data.models.Transaction
import com.ledgerly.data.repository.BudgetRepository
import kotlinx.coroutines.launch
import java.time.LocalDate

class BudgetViewModel(
    application: Application
) : AndroidViewModel(application) {

    private val repository: BudgetRepository

    private val _budgets = mutableStateListOf<Budget>()

    val budgets: List<Budget>
        get() = _budgets

    val activeBudgets: List<Budget>
        get() = _budgets.filter {
            it.limit > 0.0
        }

    val overallBudget: Budget?
        get() = _budgets.firstOrNull {
            it.id == OVERALL_BUDGET_ID
        }

    val categoryBudgets: List<Budget>
        get() = _budgets.filter {
            it.id != OVERALL_BUDGET_ID &&
                it.limit > 0.0
        }

    val hasOverallBudget: Boolean
        get() = overallBudget != null &&
            (overallBudget?.limit ?: 0.0) > 0.0

    val allocatedTotal: Double
        get() = categoryBudgets.sumOf {
            it.limit
        }

    val unallocatedAmount: Double
        get() = (overallBudget?.limit ?: 0.0) - allocatedTotal

    init {
        val database = DatabaseProvider.getDatabase(application)

        repository = BudgetRepository(
            budgetDao = database.budgetDao()
        )

        viewModelScope.launch {
            repository.budgets.collect { entities ->
                _budgets.clear()
                _budgets.addAll(
                    entities.map { entity ->
                        entity.toModel()
                    }
                )
            }
        }
    }

    fun addBudget(
        category: Category,
        limit: Double
    ) {
        setBudget(
            category = category,
            limit = limit
        )
    }

    fun setBudget(
        category: Category,
        limit: Double
    ) {
        val existingBudget = _budgets.firstOrNull {
            it.category.id == category.id
        }

        val budget = if (existingBudget == null) {
            Budget(
                id = category.id,
                category = category,
                limit = limit,
                spent = 0.0
            )
        } else {
            existingBudget.copy(
                limit = limit
            )
        }

        viewModelScope.launch {
            repository.saveBudget(
                budget.toEntity()
            )
        }
    }

    fun setOverallBudget(
        limit: Double
    ) {
        val existingBudget = overallBudget

        val budget = if (existingBudget == null) {
            Budget(
                id = OVERALL_BUDGET_ID,
                category = overallBudgetCategory,
                limit = limit,
                spent = 0.0
            )
        } else {
            existingBudget.copy(
                limit = limit
            )
        }

        viewModelScope.launch {
            repository.saveBudget(
                budget.toEntity()
            )
        }
    }

    fun saveBudgetPlan(
        overallLimit: Double,
        allocations: List<Pair<Category, Double>>
    ) {
        setOverallBudget(
            limit = overallLimit
        )

        allocations.forEach { allocation ->
            val category = allocation.first
            val limit = allocation.second

            if (limit > 0.0) {
                setBudget(
                    category = category,
                    limit = limit
                )
            }
        }
    }

    fun refreshCurrentMonthBudgets(
        transactions: List<Transaction>
    ) {
        val today = LocalDate.now()
        val startDate = today.withDayOfMonth(1)
        val endDate = today.withDayOfMonth(
            today.lengthOfMonth()
        )

        updateBudgets(
            transactions = transactions,
            startDate = startDate,
            endDate = endDate
        )
    }

    fun updateBudgets(
        transactions: List<Transaction>,
        startDate: LocalDate,
        endDate: LocalDate
    ) {
        _budgets.indices.forEach { index ->
            val budget = _budgets[index]

            val spent = if (budget.id == OVERALL_BUDGET_ID) {
                transactions
                    .filter {
                        !it.isIncome &&
                            !it.date.isBefore(startDate) &&
                            !it.date.isAfter(endDate)
                    }
                    .sumOf {
                        it.amount
                    }
            } else {
                transactions
                    .filter {
                        !it.isIncome &&
                            it.category.id == budget.category.id &&
                            !it.date.isBefore(startDate) &&
                            !it.date.isAfter(endDate)
                    }
                    .sumOf {
                        it.amount
                    }
            }

            val updatedBudget = budget.copy(
                spent = spent
            )

            _budgets[index] = updatedBudget

            viewModelScope.launch {
                repository.saveBudget(
                    updatedBudget.toEntity()
                )
            }
        }
    }

    fun budgetFor(
        category: Category
    ): Budget? {
        return _budgets.firstOrNull {
            it.category.id == category.id
        }
    }

    fun deleteBudget(
        budget: Budget
    ) {
        viewModelScope.launch {
            repository.deleteBudgetById(
                budget.id
            )
        }
    }

    fun deleteOverallBudget() {
        viewModelScope.launch {
            repository.deleteBudgetById(
                OVERALL_BUDGET_ID
            )
        }
    }

    val totalLimit: Double
        get() = overallBudget?.limit
            ?: categoryBudgets.sumOf {
                it.limit
            }

    val totalSpent: Double
        get() = overallBudget?.spent
            ?: categoryBudgets.sumOf {
                it.spent
            }

    val totalRemaining: Double
        get() = totalLimit - totalSpent

    companion object {
        const val OVERALL_BUDGET_ID = "__overall__"

        val overallBudgetCategory = Category(
            id = OVERALL_BUDGET_ID,
            name = "Overall Budget",
            systemIcon = "wallet",
            isDefault = true
        )
    }
}